wallet/cached_key_deriver_test.go

```diff
@@ -0,0 +1,404 @@
+package wallet
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+
+	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
+	"github.com/stretchr/testify/assert"
+)
+
+type MockKeyDeriver struct {
+	publicKeyCallCount        int
+	privateKeyCallCount       int
+	symmetricKeyCallCount     int
+	specificSecretCallCount   int
+	publicKeySleepTime        time.Duration
+	publicKeyToReturn         *ec.PublicKey
+	privateKeyToReturn        *ec.PrivateKey
+	symmetricKeyToReturn      *ec.SymmetricKey
+	specificSecretToReturn    []byte
+	symmetricKeyErrorToReturn error
+}
+
+func (m *MockKeyDeriver) DerivePublicKey(protocolID Protocol, keyID string, counterparty Counterparty, forSelf bool) (*ec.PublicKey, error) {
+	if m.publicKeySleepTime > 0 {
+		time.Sleep(m.publicKeySleepTime)
+	}
+	m.publicKeyCallCount++
+	return m.publicKeyToReturn, nil
+}
+
+func (m *MockKeyDeriver) DerivePrivateKey(protocolID Protocol, keyID string, counterparty Counterparty) (*ec.PrivateKey, error) {
+	m.privateKeyCallCount++
+	return m.privateKeyToReturn, nil
+}
+
+func (m *MockKeyDeriver) DeriveSymmetricKey(protocolID Protocol, keyID string, counterparty Counterparty) (*ec.SymmetricKey, error) {
+	m.symmetricKeyCallCount++
+	return m.symmetricKeyToReturn, m.symmetricKeyErrorToReturn
+}
+func (m *MockKeyDeriver) RevealSpecificSecret(counterparty Counterparty, protocol Protocol, keyID string) ([]byte, error) {
+	m.specificSecretCallCount++
+	return m.specificSecretToReturn, nil
+}
+
+func TestDerivePublicKey(t *testing.T) {
+	// Create keys and cached key deriver
+	rootKey, _ := ec.PrivateKeyFromBytes([]byte{1})
+	publicKey := &ec.PublicKey{X: big.NewInt(0), Y: big.NewInt(0), Curve: ec.S256()}
+
+	// Create parameters
+	protocol := Protocol{
+		SecurityLevel: SecurityLevelSilent,
+		Protocol:      "testprotocol",
+	}
+	keyID := "key1"
+	counterparty := Counterparty{
+		Type: CounterpartyTypeSelf,
+	}
+
+	t.Run("should call derivePublicKey on KeyDeriver and cache the result", func(t *testing.T) {
+		// Create a mock key deriver that returns a fixed public key
+		cachedDeriver := NewCachedKeyDeriver(rootKey, 0)
+		mockKeyDeriver := &MockKeyDeriver{publicKeyToReturn: publicKey}
+		cachedDeriver.keyDeriver = mockKeyDeriver
+
+		// First call - should call through to real deriver
+		pubKey1, err := cachedDeriver.DerivePublicKey(protocol, keyID, counterparty, false)
+		assert.NoError(t, err)
+		assert.NotNil(t, pubKey1)
+		assert.Equal(t, publicKey.ToDERHex(), pubKey1.ToDERHex())
+
+		// Second call - should return cached value
+		pubKey2, err := cachedDeriver.DerivePublicKey(protocol, keyID, counterparty, false)
+		assert.NoError(t, err)
+		assert.Equal(t, pubKey1.ToDERHex(), pubKey2.ToDERHex())
+		assert.Equal(t, mockKeyDeriver.publicKeyCallCount, 1)
```

Comment:
consider using adding a message to all those assert, to give in assertion failure, some hint on which call it failed.

for example
```go
assert.NoErrorf(t, err, "first call to DerivePublicKey")
//....

assert.NoErrorf(t, err, "second call to DerivePublicKey")
//...
```


similar for the tests below

Reply:
And I have similar thoughts for the tests below